### PR TITLE
Ported over HaveIBeenPwned from focus.

### PR DESCRIPTION
**Pull request by eskimor**

### CODE REVIEW

File: backend/Rhyolite/Backend/HaveIBeenPwned.hs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  = HaveIBeenPwnedResult_Undisclosed
+  | HaveIBeenPwnedResult_Disclosed Int
+  | HaveIBeenPwnedResult_ApiError
+  deriving (Eq, Ord, Show)
+
+class Monad m => MonadPwned m where
+  -- | returns the number of disclosures the supplied password has been seen in.
+  --
+  -- if this is not zero, do not use the supplied password, it is known to hackers.
+  -- if it is *is* zero, it might still not be safe, only that if it is
```
Comment (alexfmpe) [inline backend/Rhyolite/Backend/HaveIBeenPwned.hs]: duplicate 'is'

File: backend/Rhyolite/Backend/HaveIBeenPwned.hs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  -- compramised, that is not yet known
```
Comment (alexfmpe) [inline backend/Rhyolite/Backend/HaveIBeenPwned.hs]: 'compromised'

File: backend/Rhyolite/Backend/HaveIBeenPwned.hs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      return HaveIBeenPwnedResult_ApiError
+    Right result -> case responseStatus result of
+      Status 200 _ -> return $ case parseHIBPResponse (responseBody result) rest of
+        0 -> HaveIBeenPwnedResult_Undisclosed
+        n -> HaveIBeenPwnedResult_Disclosed n
+      Status code phrase -> do
+        $(logError) $ T.pack $ show $ Status code phrase
+        return HaveIBeenPwnedResult_ApiError
+
+
```
Comment (alexfmpe) [inline backend/Rhyolite/Backend/HaveIBeenPwned.hs]: duplicate 'two'

File: backend/Rhyolite/Backend/HaveIBeenPwned.hs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+-- | get the sha1 digest for the supplied password, split into two two parts, to agree with the
+--   hibp api
+passwdDigest :: Text -> (Text, Text)
+passwdDigest passwd = (T.take 5 digest, T.drop 5 digest)
+  where digest = T.toUpper $ T.pack $ show $ sha1 $ encodeUtf8 passwd
+        sha1 :: ByteString -> Digest SHA1
+        sha1 = hash
+
+-- | the hibp response is a line separated list of colon separated hash
+-- *suffixes* and a number indicationg the number of times that password(hash)
```
Comment (alexfmpe) [inline backend/Rhyolite/Backend/HaveIBeenPwned.hs]: 'indicating'

File: backend/Rhyolite/Backend/HaveIBeenPwned.hs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+-- has been seen in known publicly disclosed leaks
+
+parseHIBPResponse :: LBS.ByteString -> Text -> Int
+parseHIBPResponse response suffix =
+  let
+    digests :: [(LT.Text, Int)]
+    digests = fmap (fmap (read . LT.unpack . LT.drop 1) . LT.breakOn ":") $ LT.lines $ Data.Text.Lazy.Encoding.decodeUtf8 response
```
Comment (alexfmpe) [inline backend/Rhyolite/Backend/HaveIBeenPwned.hs]: should this use 'readMay' or similar?

File: backend/Rhyolite/Backend/HaveIBeenPwned.hs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  cfg <- PwnedT ask
+  let request = parseRequest_ $ T.unpack $ T.concat [_haveIBeenPwnedConfig_apihost cfg, "/", pfx]
+  result' <- liftIO $ try $ httpLbs request (_haveIBeenPwnedConfig_manager cfg)
+  case result' of
+    Left err -> do
+      $(logError) $ T.pack $ show @ HttpException $ err
+      return HaveIBeenPwnedResult_ApiError
+    Right result -> case responseStatus result of
+      Status 200 _ -> return $ case parseHIBPResponse (responseBody result) rest of
+        0 -> HaveIBeenPwnedResult_Undisclosed
```
Comment (alexfmpe) [inline backend/Rhyolite/Backend/HaveIBeenPwned.hs]: `HaveIBeenPwnedResult_Disclosed 0` is weird. Should we make this into a smart constructor and only export that?

**Comment by eskimor** [comment]
Haha, thanks @alexfmpe ! Will look into those.

File: backend/Rhyolite/Backend/HaveIBeenPwned.hs (hunk 5 above)
Comment (3noch) [reply backend/Rhyolite/Backend/HaveIBeenPwned.hs]: Yeah `read` is partial. `readMaybe` is in base.

File: backend/Rhyolite/Backend/HaveIBeenPwned.hs (hunk 6 above)
Comment (eskimor) [reply backend/Rhyolite/Backend/HaveIBeenPwned.hs]: Good point.

**Comment by ali-abrar** [comment]
Apart from `consoleHaveIBeenPwned`, which is just for testing, it looks like this module doesn't actually rely on anything else in rhyolite. We can put it here for now, but please make a task to split this out into a new package.

**Comment by ali-abrar** [comment]
Can `consoleHaveIBeenPwned` be implemented without any rhyolite deps?

**Comment by eskimor** [comment]
All in favor of moving this to its own package. The rhyolite depedencies should be easy to remove.

**Comment by eskimor** [comment]
where should this package go? obsidiansystems/haveibeenpwned ?

**Comment by ali-abrar** [comment]
https://github.com/obsidiansystems/haveibeenpwned

**Pull request closed without merging**